### DISCUSSION TITLE
Issue with filter - Postgres CDC realtime

### DISCUSSION THREAD

**Discussion by MaximeLemolt** [supabase/supabase#29884] [Questions] [answered]
I have a problem when I try to listen to UPDATE changes with Supabase realtime.

Realtime logs error message :
> "UnableToSubscribeToPostgres: {:error,\n \"Unable to subscribe to changes with given parameters. An exception happened so please check your connect parameters: [event: UPDATE, filter: id=eq.6100895b-7991-4f12-bed5-41708fa06176, schema: public, table: Player]. Exception: ERROR P0001 (raise_exception) invalid column for filter id\"}"

My code :
```
const channel = realtime.channel(`player-updates`);
channel.on(
  "postgres_changes",
  {
    event: "UPDATE",
    schema: "public",
    table: "Player",
    filter: `id=eq.${player.id}`,
  },
  (payload) => {
    console.log(payload);
  }
);

channel.subscribe(async (status) => {
  if (status === "SUBSCRIBED") {
    console.log("Ready to receive updates");
  }
});
```

When I don't specify a filter, everything works fine, but as soon as I add a filter, it no longer works, regardless of the column.
The column name is correct, everything looks OK, but it doesn't work.
If anyone has encountered this and has a solution, please let me know.
Thanks in advance.

**Comment by GaryAustin1** [comment]
Don't put your string in quotes at all.

https://supabase.com/docs/guides/realtime/postgres-changes#equal-to-eq

**Reply by MaximeLemolt** [reply]
@GaryAustin1 it was a test to solve the problem. It's not working even without the quotes.
I edited my message, my bad.
Thanks for your answer.

**Reply by GaryAustin1** [reply]
Do you have a column called id in that table?  Note your error says table Player, but your code says table Post....

Can you post  your real code and error and an image of the column definition for that table?

Also have you changed the column name or type since it was created?

**Reply by MaximeLemolt** [reply]
@GaryAustin1 

Error :

> "UnableToSubscribeToPostgres: {:error,\n \"Unable to subscribe to changes with given parameters. An exception happened so please check your connect parameters: [event: UPDATE, filter: id=eq.6100895b-7991-4f12-bed5-41708fa06176, schema: public, table: Player]. Exception: ERROR P0001 (raise_exception) invalid column for filter id\"}"

Screenshot of  the table definition :
<img width="1434" alt="player_table_id_column" src="https://github.com/user-attachments/assets/493138cc-839b-4122-8ed4-f69f1c93d7f6">

I I've never changed the column type or definition.

Thanks

**Reply by GaryAustin1** [reply]
Seems right.
Can you do a quick test with just a text value (not a uuid string) to see if it gets past the error?  It should not care what the string looks like but....

**Reply by GaryAustin1** [reply]
If that does not work I would try restarting the instance.
Also sometimes users with 2 instances are going to the wrong instance but tables are similar but not the same.  So the UI is not matching the info the call is seeing.

Otherwise I've got nothing else.  I see you also have an issue generated for them in the realtime repository.  You should also fix the typos in that or that will likely be the first response to it.

**Reply by MaximeLemolt** [reply]
@GaryAustin1 

> Can you do a quick test with just a text value (not a uuid string) to see if it gets past the error? It should not care what the string looks like but....

II've already done this test and the error is still present, I've also tested all the columns in the table and got the same result

> If that does not work I would try restarting the instance.
Also sometimes users with 2 instances are going to the wrong instance but tables are similar but not the same. So the UI is not matching the info the call is seeing.

I've already restarted the project on supabase several times. Maybe it caused some inconsistencies.

**Reply by MaximeLemolt** [reply]
@GaryAustin1 
I'll try to create a new project on supabase to see if that solves the problem.
The problem occurred all at once, by which I mean that using the filters was working and when I resumed development of the application this morning it was no longer working.

I ran into a lot of problems with SQL connections, which led me to reset my migrations (I use Prisma for that). Maybe it finally broke everything on the supabase side.

**Reply by GaryAustin1** [reply]
I notice you seem to be using Prisma (based on your table image).
There are issues with realtime and Prisma, but they are usually caused by grants being wiped out by Prisma.  The filter error though would not be what I would expect and you said it worked without the filter and that case I would think would also error.
Troubleshooting in this guide covers that: https://supabase.com/partners/integrations/prisma

**Reply by MaximeLemolt** [reply, marked as answer]
I solved the problem by creating a new supabase project. The filters work again without having modified the code.
I think the multiple project restarts caused problems.

I have the impression that the problem was perhaps linked to the RLS, as I notice the presence of a padlock icon next to the table name, which was no longer present on the old project despite the deactivation of the RLS.

**Old project screenshot (with the error) :**

<img width="1434" alt="player_table_id_column" src="https://github.com/user-attachments/assets/8656f205-7c11-4165-8379-0214e1c53a8f">


**New project screenshot (without the error) :**

<img width="575" alt="Capture d’écran 2024-10-15 à 13 44 02" src="https://github.com/user-attachments/assets/daa48e9c-1b0a-43eb-8e3e-9c5b52d15005">



@GaryAustin1 thanks for your help 👍

**Comment by Rubenburdin** [comment]
For anyone else hitting this "invalid column for filter" error, few things to check before nuking the project:
Prisma specific: Prisma migrations can wipe grants on tables. Realtime needs specific permissions to read from the replication stream. Run this to restore them:
`grant select on your_table to supabase_realtime_admin;`

Also make sure your table is in the supabase_realtime publication:
`alter publication supabase_realtime add table your_table;`

RLS gotcha: Even with RLS "disabled" in the UI, sometimes the policies stick around in a weird state. The padlock icon difference you noticed is telling. You can verify with:
`select relname, relrowsecurity 
from pg_class 
where relname = 'Player';`

If relrowsecurity is true but you expect it off, run alter table "Player" disable row level security;
Filter column must be in replica identity: By default Postgres only includes primary key columns in the WAL for updates. If you're filtering on a non PK column, you need:
`alter table "Player" replica identity full;`


Otherwise Realtime can't see that column's value in the change event to apply your filter.
The "create new project" fix likely just reset all these to defaults. Worth checking the above if migrating isn't an option.
